.github/workflows/checks.yaml

```diff
@@ -0,0 +1,12 @@
+name: checks
```

Comment:
Actually, I usually prefer having them in a single file which is usually named `build.yaml`. [Here](https://github.com/umutoztunc/whitesymex/blob/main/.github/workflows/build.yaml) is an example. The reason for doing it like this is the fact that this creates a single docker, let's say `ubuntu-latest`, and does every step in the given order. If a step fails, it does not continue and exits failed with the result of the steps till the failed one.

However, if you split this single job to multiple jobs, they run in parallel. Let's say that you have jobs for `linting`, `formatting check`, `building` and `testing`. If the `linter` fails, this means that the code won't compile and it is not needed to do style checks or try to build it. Same goes for `testing` and `building`: `testing` depends on `building` job to be completed first. In this case, you need to chain these jobs with dependencies which means that they are not working in parallel anymore and they are recreating an `ubuntu-latest` instance again and again for no reason. Therefore, I only split them into jobs when they are independent but logically related jobs which separates them and also makes it run faster since they can run in parallel.

Splitting the steps to multiple files would result in the same issues as above. Therefore, I think we should split things when they are not related. I am not sure about the file name, but I think we will have the steps:

1. Check syntax errors (possibly with clang-tidy).
2. Check code style and formatting if not already done in the above step (using clang-format).
3. Build and run the tests (possibly using gtest).

Reply:
Thanks for the thorough explanations.
BTW, we might need to build the test before clang-tidy, so that we can provide it a compilation database.